File: app/ios/ebolarisk/LogRegModel.swift
//
//  LogRegModel.swift
//  ebolarisk
//
//  Created by Andres Colubri on 4/14/18.
//  Copyright © 2018 broadinstitute. All rights reserved.
//

import Foundation

class LogRegModel {
    var intercept: Float
    var rangeScale: Float
    var terms: [String: ModelTerm]
    
    init(modelName name: String) {
        intercept = 0
        rangeScale = 0
        terms = [:]
        
        loadTerms(name: name)
        loadMinMax(name: name)
    }
    
    func matches(vars: Array<String>) -> Bool {
        let mvars = Array(terms.keys)
        return contains(first: mvars, second: vars) && contains(first: vars, second: mvars)
    }
    
    func contains(vars: Array<String>) -> Bool {
        let mvars = Array(terms.keys)
        return contains(first: vars, second: mvars)
    }
    
    func setIntercept(b0: Float) {
        intercept = b0;
    }
    
    func addTerm(term: ModelTerm) {
        terms[term.varName] = term
    }
    
    func eval(values: [String: Float]) -> Float {
        return sigmoid(z: evalScore(values: values))
    }

    func evalScore(values: [String: Float]) -> Float {
        var score = intercept
        for varName in values.keys {
            if let t = terms[varName], let x = values[varName] {
                score += t.eval(x: x);
            }
        }
        return score
    }
 
    func evalDetails(values: [String:Float]) -> [String:Array<Float>] {
        var details = [String:Array<Float>]()
        for varName in values.keys {
            if let t = terms[varName], let v = values[varName] {
                let coeff0 = t.coeff(i: 0)
                let contrib = t.eval(x: v)
                let scaledRange = abs(t.max - t.min) / rangeScale
                let scaledContrib = min(abs(contrib - t.min) / rangeScale, scaledRange)
                details[varName] = [v, contrib, scaledContrib, scaledRange, coeff0]                
            }
        }
        return details
    }
    
    func loadTerms(name: String) {
//        print("LOADING TERMS")
        let lines = loadStrings(forResource: "mice", ofType: "txt", inDirectory: "models/" + name)
        if let pos = indexOf(source: lines[0], substring: "est", offset: 2) {
            var rcsCoeffs = [Float]()
            let parenthesis = CharacterSet.init(charactersIn: "()")
            let apostrophe = CharacterSet.init(charactersIn: "'")
            
            var n = 1
            while (true) {
                let line = lines[n]
                n += 1
                if line.endIndex < pos {
                    continue
                }
                
                let s = trim(source: line[..<pos])
                let v = s.split(separator: " ")
                
                if v.count == 1 {
                    break
                }
                
                let valueStr = v[v.count - 1]
                if let value = Float(valueStr) {
                    if let pos0 = indexOf(source: s, substring: String(valueStr), offset: 0) {
                        let varStr = trim(source: s[..<pos0])
                        if startsWidth(source: varStr, substring: "rcs") {
                            if let pos1 = lastIndexOf(source: varStr, substring: ")", offset: 0) {
                                let rcsString = varStr[varStr.index(varStr.startIndex, offsetBy: 4)..<pos1]
                                let pieces = rcsString.split(separator: "c")

                                let part1 = pieces[0].split(separator: ",")
                                let varName = trim(source: part1[0])
                                if let rcsOrder = Int(trim(source: part1[1])) {
                                    let knotStr = pieces[1].trimmingCharacters(in: parenthesis).split(separator: ",")
                                    var rcsKnots = [Float]()
                                    for v in knotStr {
                                        rcsKnots.append(Float(trim(source: v))!)
                                    }
                                    
                                    let coeffOrder = varStr.count - varStr.trimmingCharacters(in: apostrophe).count
                                    
                                    if coeffOrder == 0 {
                                        rcsCoeffs = Array(repeating: 0.0, count: rcsOrder - 1)
                                    }
                                    
                                    if 0 < rcsCoeffs.count {
                                        rcsCoeffs[coeffOrder] = value
                                    }
                                    
                                    if coeffOrder == rcsOrder - 2 {
//                                        print("Adding RCS term of order " + String(rcsOrder) + " for " + varName + " with the following coefficients and knots:")
//                                        print(rcsCoeffs)
//                                        print(rcsKnots)
                                        let rcs = RCSTerm(variableName: varName, rcsOrder: rcsOrder, coefficients: rcsCoeffs, knots: rcsKnots)
                                        addTerm(term: rcs)
                                    }
                                }
                            }
                        } else {
                            if startsWidth(source: varStr, substring: "(Intercept)") {
//                                print("Setting the intercept to " + String(value))
                                setIntercept(b0: value);
                            } else {
//                                print("Adding linear term for " + varStr + " with the following coefficients: " + String(value))
                                let lin = LinearTerm(variableName: varStr, coefficient: value)
                                addTerm(term: lin);
                            }
                        }
                    }
                }
            }
        }
    }
    
    func loadMinMax(name: String) {
//        print("LOADING MINMAX")
        let lines = loadStrings(forResource: "minmax", ofType: "txt", inDirectory: "models/" + name)
        rangeScale = 0
        for line in lines {
            let pieces = line.split(separator: " ")
            if pieces.count < 3 {
                continue
            }
            let name = String(pieces[0])
            if let term = terms[name], let min = Float(pieces[1]), let max = Float(pieces[2]) {
//                print("Setting min and max for " + name + " " + String(min) + " " + String(max))
                term.setMin(m: min)
                term.setMax(m: max)
                let f = abs(max - min)
                if rangeScale < f {
                    rangeScale = f
//                    print("Updating range scale to " + String(rangeScale))
                }
            }
        }
    }
    
    func loadStrings(forResource name: String?, ofType ext: String?, inDirectory subpath: String?) -> Array<String> {
        //  https://stackoverflow.com/questions/27206176/where-to-place-a-txt-file-and-read-from-it-in-a-ios-project
        if let path = Bundle.main.path(forResource: name, ofType: ext, inDirectory: subpath) {
            let fm = FileManager()
            let exists = fm.fileExists(atPath: path)
            if exists {
                let content = fm.contents(atPath: path)
                if let contentAsString = String(data: content!, encoding: String.Encoding.utf8) {
                    let lines = contentAsString.components(separatedBy: "\n")
                    return lines
                }
            }
        }
        return [""]
    }
    
    // Utils
    
    func sigmoid(z: Float) -> Float {
        return 1.0 / (1.0 + exp(-z))
    }
    
    func contains(first: Array<String>, second: Array<String>) -> Bool {
        var all = true
        for element in second {
            if !first.contains(element) {
                all = false
                break
            }
        }
        return all
    }
    
    // String API cheatsheet
    // https://useyourloaf.com/blog/swift-string-cheat-sheet/
    func indexOf(source: String, substring: String, offset: Int) -> String.Index? {
        if let range = source.range(of: substring) {
            return source.index(range.lowerBound, offsetBy: offset, limitedBy: source.endIndex)
        } else {
            return nil
        }
    }
    
    func lastIndexOf(source: String, substring: String, offset: Int) -> String.Index? {
        if let range = source.range(of: substring, options: String.CompareOptions.backwards) {
            return source.index(range.lowerBound, offsetBy: offset, limitedBy: source.endIndex)
        } else {
            return nil
        }
    }
    
    func startsWidth(source: String, substring: String) -> Bool {
        if let range = source.range(of: substring) {
            return range.lowerBound == source.startIndex
        } else {
            return false
        }
    }
    
    func trim(source: String) -> String {
        return source.trimmingCharacters(in: .whitespacesAndNewlines)
    }
    
    func trim(source: Substring.SubSequence) -> String {
        return source.trimmingCharacters(in: .whitespacesAndNewlines)
    }
    
    func toString() -> String {
        var res = "Intercept: " + String(intercept)
        for term in terms.values {
            res += "\n" + term.toString();
        }
        return res;
    }
    
    class ModelTerm {
        var varName: String
        var min, max: Float
        
        init(variableName name: String) {
            varName = name
            min = 0
            max = 0
        }
        
        func coeff(i: Int) -> Float { return 0 }
        func eval(x: Float) -> Float { return 0 }
        func setMin(m: Float) { min = m }
        func setMax(m: Float) { max = m }
        func toString() -> String { return "" }
    }
    
    class LinearTerm: ModelTerm {
        var coeff: Float
        
        init(variableName name: String, coefficient c: Float) {
            coeff = c
            super.init(variableName: name)
        }
        
        override func coeff(i: Int) -> Float { return coeff }
        override func eval(x: Float) -> Float { return coeff * x }
        
        override func toString() -> String {
            var res = "Linear term for " + varName + "\n"
            res += "  Coefficient: " + String(coeff)
            res += "\n  Minimum value: " + String(min)
            res += "\n  Maximum value: " + String(max)
            return res
        }
    }
    
    class RCSTerm: ModelTerm {
        var order: Int
        var coeffs: Array<Float>
        var knots: Array<Float>

        init(variableName name: String, rcsOrder k: Int, coefficients c: Array<Float>, knots kn: Array<Float>) {
            order = k
            coeffs = c
            knots = kn
            super.init(variableName: name)
        }
        
        override func coeff(i: Int) -> Float {
            return coeffs[i]
        }
        
        override func eval(x: Float) -> Float {
            var sum = coeffs[0] * x
            for t in 1 ... order - 2 {
                sum += coeffs[t] * rcs(x: x, term: t);
            }
            return sum;
        }
        
        func cubic(u: Float) -> Float {
            let t = u > 0 ? u : 0
            return t * t * t
        }
        
        func rcs(x: Float, term: Int) -> Float {
            let k = order - 1
            let j = term - 1
            let t = knots
            let c = (t[k] - t[0]) * (t[k] - t[0])
            let value = +cubic(u: x - t[j]) - cubic(u: x - t[k - 1]) * (t[k] - t[j]) / (t[k] - t[k-1]) + cubic(u: x - t[k]) * (t[k - 1] - t[j]) / (t[k] - t[k-1])
            return value / c
        }
        
        override func toString() -> String {
            var res = "RCS term of order " + String(order) + " for " + varName + "\n"
            res += "  Coefficients:"
            for i in 0 ... coeffs.count - 1 {
                res += " " + String(coeffs[i])
            }
            res += "\n"
            res += "  Knots:"
            for i in 0 ... knots.count - 1 {
                res += " " + String(knots[i])
            }
            res += "\n  Minimum value: " + String(min)
            res += "\n  Maximum value: " + String(max)
            return res
        }
    }
}

File: app/android/app/src/main/java/org/broadinstitute/ebolarisk/LogRegModel.java
package org.broadinstitute.ebolarisk;

import org.apache.commons.math3.analysis.function.Sigmoid;
import processing.core.PApplet;

import java.io.InputStream;
import java.util.HashMap;
import java.util.Set;

public class LogRegModel {
  protected float intercept;
  protected float rangeScale;
  protected Sigmoid sigmoid;
  protected HashMap<String, ModelTerm> terms;

  public LogRegModel(InputStream modelIS, InputStream minmaxIS) {
    intercept = 0;
    terms = new HashMap<String, ModelTerm>();
    sigmoid = new Sigmoid();

    loadTerms(modelIS);
    loadMinMax(minmaxIS);
  }

  public boolean matches(Set<String> vars) {
    Set<String> mvars = terms.keySet();
    return vars.containsAll(mvars) && mvars.containsAll(vars);
  }

  public boolean contains(Set<String> vars) {
    Set<String> mvars = terms.keySet();
    return vars.containsAll(mvars);
  }

  void setIntercept(float b0) {
    intercept = b0;
  }

  void addTerm(ModelTerm term) {
    terms.put(term.varName, term);
  }

  public float eval(HashMap<String, Float> values) {
    return (float)sigmoid.value(evalScore(values));
  }

  public float evalScore(HashMap<String, Float> values) {
    float score = intercept;
    for (String var: values.keySet()) {
      ModelTerm t = terms.get(var);
      if (t == null) continue;
      score += t.eval(values.get(var));
    }
    return score;
  }

  public HashMap<String, Float[]> evalDetails(HashMap<String, Float> values) {
    HashMap<String, Float[]> details = new HashMap<String, Float[]>();
    for (String var: values.keySet()) {
      ModelTerm t = terms.get(var);
      if (t == null) continue;
      float value = values.get(var);
      float coeff0 = t.coeff(0);
      float contrib = t.eval(value);
      float scaledRange = PApplet.abs(t.max - t.min) / rangeScale;
      float scaledContrib = PApplet.min(PApplet.abs(contrib - t.min) / rangeScale, scaledRange);
      details.put(t.varName, new Float[]{value, contrib, scaledContrib, scaledRange, coeff0});
    }
    return details;
  }

  void loadTerms(InputStream is) {
    float[] rcsCoeffs = null;
    String[] lines = PApplet.loadStrings(is);

    int pos = lines[0].indexOf("est") + 2;

    int n = 1;
    while (true) {
      String line = lines[n++];
      String s = line.substring(0, pos).trim();
      String[] v = s.split(" ");
      if (v.length == 1) break;
      String valueStr = v[v.length - 1];
      float value = PApplet.parseFloat(valueStr);

      int pos0 = s.indexOf(valueStr);
      String varStr = s.substring(0, pos0).trim();

      if (varStr.indexOf("rcs") == 0) {
        int pos1 = varStr.lastIndexOf(")");
        String rcsString = varStr.substring(4, pos1);
        String[] pieces = rcsString.split("c");

        String[] part1 = pieces[0].split(",");
        String varName = part1[0].trim();
        int rcsOrder = PApplet.parseInt(part1[1].trim());
        String[] knotStr = pieces[1].replace("(", "").replace(")", "").split(",");
        float[] rcsKnots = new float[knotStr.length];
        for (int i = 0; i < knotStr.length; i++) {
          rcsKnots[i] = PApplet.parseFloat(knotStr[i]);
        }

        int coeffOrder = varStr.length() - varStr.replace("'", "").length();
        if (coeffOrder == 0) {
          rcsCoeffs = new float[rcsOrder - 1];
        }
        if (rcsCoeffs != null) rcsCoeffs[coeffOrder] = value;

        if (coeffOrder == rcsOrder - 2) {
          ModelTerm rcs = new RCSTerm(varName, rcsOrder, rcsCoeffs, rcsKnots);
          addTerm(rcs);
        }
      } else {
        if (varStr.equals("(Intercept)")) {
          setIntercept(value);
        } else {
          ModelTerm lin = new LinearTerm(varStr, value);
          addTerm(lin);
        }
      }
    }
  }

  void loadMinMax(InputStream is) {
    String[] lines = PApplet.loadStrings(is);
    rangeScale = 0;
    for (String line: lines) {
      String[] pieces = line.split(" ");
      String name = pieces[0];
      float min = PApplet.parseFloat(pieces[1]);
      float max = PApplet.parseFloat(pieces[2]);
      ModelTerm term = terms.get(name);
      if (term != null) {
        term.setMin(min);
        term.setMax(max);
        float f = PApplet.abs(max - min);
        if (rangeScale < f) rangeScale = f;
      }
    }
  }

  public String toString() {
    String res = "Intercept: " + intercept;
    for (ModelTerm term: terms.values()) {
      res += "\n" + term.toString();
    }
    return res;
  }

  class ModelTerm {
    String varName;
    float min, max;

    ModelTerm(String name) { varName = name; }
    float coeff(int i) { return 0; }
    float eval(float x) { return 0; }
    void setMin(float m) { min = m; }
    void setMax(float m) { max = m; }
  }

  class LinearTerm extends ModelTerm {
    float coeff;

    LinearTerm(String name, float c) {
      super(name);
      coeff = c;
    }
    float coeff(int i) { return coeff; }
    float eval(float x) {
      return coeff * x;
    }
    public String toString() {
      String res = "Linear term for " + varName + "\n";
      res += "  Coefficient: " + coeff;
      res += "\n  Minimum value: " + min;
      res += "\n  Maximum value: " + max;
      return res;
    }
  }

  class RCSTerm extends ModelTerm {
    int order;
    float[] coeffs;
    float[] knots;

    RCSTerm(String name, int k, float[] c, float[] kn) {
      super(name);
      order = k;
      coeffs = new float[k - 1];
      knots = new float[k];
      PApplet.arrayCopy(c, coeffs);
      PApplet.arrayCopy(kn, knots);
    }

    float cubic(float u) {
      float t = u > 0 ? u : 0;
      return t * t * t;
    }

    float rcs(float x, int term) {
      int k = order - 1;
      int j = term - 1;
      float[] t = knots;
      float c = (t[k] - t[0]) * (t[k] - t[0]);
      float value = +cubic(x - t[j])
                    -cubic(x - t[k - 1]) * (t[k] - t[j])/(t[k] - t[k-1])
                    +cubic(x - t[k]) * (t[k - 1] - t[j])/(t[k] - t[k-1]);
      return value / c;
    }

    float coeff(int i) {
      return coeffs[i];
    }

    float eval(float x) {
      float sum = coeffs[0] * x;
      for (int t = 1; t < order - 1; t++) {
        sum += coeffs[t] * rcs(x, t);
      }
      return sum;
    }

    public String toString() {
      String res = "RCS term of order " + order + " for " + varName + "\n";
      res += "  Coefficients:";
      for (int i = 0; i < coeffs.length; i++) {
        res += " " + coeffs[i];
      }
      res += "\n";
      res += "  Knots:";
      for (int i = 0; i < knots.length; i++) {
        res += " " + knots[i];
      }
      res += "\n  Minimum value: " + min;
      res += "\n  Maximum value: " + max;
      return res;
    }
  }
}
